File: python/opall.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2013, blarz <simon@blarzwurst.de>
#
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

SCRIPT_NAME    = 'opall'
SCRIPT_AUTHOR  = 'blarz'
SCRIPT_VERSION = '0.1'
SCRIPT_LICENSE = 'ISC'
SCRIPT_DESC    = 'Give op to everybody, like /op -Yes * in irssi'

try:
    import weechat
    import_ok = True
except ImportError:
    print('This script must be run under WeeChat.')
    print('Get WeeChat now at: http://www.weechat.org/')
    import_ok = False

def withoutOp(server, channel):
    L = []

    infolist = weechat.infolist_get('irc_nick', '', '%s,%s' %(server, channel))
    while weechat.infolist_next(infolist):
        if not '@' in weechat.infolist_string(infolist, 'prefix'):
            L.append(weechat.infolist_string(infolist, 'name'))
    weechat.infolist_free(infolist)

    return L

def opall(data, buffer, args):
    channel = weechat.buffer_get_string(buffer, 'localvar_channel')
    server = weechat.buffer_get_string(buffer, 'localvar_server')

    if not weechat.info_get('irc_is_channel', channel):
        weechat.prnt(buffer, '%sopall: Not an IRC channel' % weechat.prefix('error'))
        return weechat.WEECHAT_RC_OK

    toOp = withoutOp(server, channel)
    if len(toOp) == 0:
        return weechat.WEECHAT_RC_OK

    # how many people can we op at once
    modes = int(weechat.info_get('irc_server_isupport_value', '%s,MODES' % server)) or 0
    if modes == 0:
        weechat.prnt(buffer, '%sopall: failed to determine MODES' % weechat.prefix('error'))
        return weechat.WEECHAT_RC_ERROR

    frm = 0
    to = modes
    while len(toOp) > frm:
        weechat.command(buffer, '/OP %s' % ' '.join(toOp[frm:to]))
        frm = to
        to += modes

    return weechat.WEECHAT_RC_OK

if __name__ == '__main__' and import_ok:
    weechat.register(SCRIPT_NAME, SCRIPT_AUTHOR, SCRIPT_VERSION, SCRIPT_LICENSE, SCRIPT_DESC, '', '')

    weechat.hook_command(SCRIPT_NAME, SCRIPT_DESC, '', '', '', 'opall', '');

File: javascript/opall.js
name = "opall";
author = "gagz@riseup.net";
version = "0.2";
license = "wtfpl";
description = "op people using chanserv instead of /mode";
shutdown_function = "";
charset = "";

weechat.register(name, author, version, license, description, shutdown_function, charset);

weechat.hook_command("opall", "OP everybody on the channel, using chanserv instead of /mode", "", "", "", "chanserv_op_all", "");

function chanserv_op_all() {
        var buffer = weechat.current_buffer()
        var chan = weechat.buffer_get_string(buffer, "localvar_channel")

        // we must be sure to be on an IRC buffer
        if( weechat.buffer_get_string(buffer, "plugin") != "irc" ) {
                weechat.print("", "Works only on IRC channels")
                return weechat.WEECHAT_RC_ERROR
        }

        // lets get the nicklist of the current buffer
        var nicklist = weechat.infolist_get("nicklist", buffer, "");
        // and walk through it
        while( weechat.infolist_next(nicklist) ) {
                var type = weechat.infolist_string(nicklist, "type");
                var visible = weechat.infolist_integer(nicklist, "visible");
                var prefix = weechat.infolist_string(nicklist, "prefix");

                // we are only interested in actual non-op visible nicks
		// TODO: find a more reliable way to op non-op users (ie. prefix
		// can be changed in the settings and might not be "@")
		// TODO: check the IRC server/services version to talk with
		// chanserv correctly. This works with charybdis/atheme.
                if( type == "nick" && visible == 1 && prefix != "@") {
                        var nick = weechat.infolist_string(nicklist, "name");
                        var command = "/msg chanserv op " + chan + " " + nick;
                        weechat.print("", command);
                        weechat.command(buffer, command);
                }
        }
        weechat.infolist_free(nicklist);
        return weechat.WEECHAT_RC_OK;
}
